Fix invalid special casing of the unreachable! macro

Urgau opened the pull request:
This pull-request fix an invalid special casing of the `unreachable!` macro in the same way the `panic!` macro was solved, by adding two new internal only macros `unreachable_2015` and `unreachable_2021` edition dependent and turn `unreachable!` into a built-in macro that do dispatching. This logic is stolen from the `panic!` macro.

~~This pull-request also adds an internal feature `format_args_capture_non_literal` that allows capturing arguments from formatted string that expanded from macros. The original RFC #2795 mentioned this as a future possibility. This feature is [required](https://github.com/rust-lang/rust/issues/92137#issuecomment-1018630522) because of concatenation that needs to be done inside the macro:~~
```rust
$crate::concat!("internal error: entered unreachable code: ", $fmt)
```

**In summary** the new behavior for the `unreachable!` macro with this pr is:

Edition 2021:
```rust
let x = 5;
unreachable!("x is {x}");
```
```
internal error: entered unreachable code: x is 5
```

Edition <= 2018:
```rust
let x = 5;
unreachable!("x is {x}");
```
```
internal error: entered unreachable code: x is {x}
```

Also note that the change in this PR are **insta-stable** and **breaking changes** but this a considered as being a [bug](https://github.com/rust-lang/rust/issues/92137#issuecomment-998441613).
If someone could start a perf run and then a crater run this would be appreciated.

Fixes https://github.com/rust-lang/rust/issues/92137

Urgau committed:
Fix invalid special casing of the unreachable! macro

rust-highfive commented:
r? @oli-obk

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4903351949?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/a48ee089f9b8a9acd38fd974f0167eead5ae7244/checks/4903351949/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Some tests failed in compiletest suite=mir-opt mode=mir-opt host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
......F.........................i...............................
failures:

---- [mir-opt] mir-opt/issue_76432.rs stdout ----
21       let mut _19: *const T;               // in scope 0 at $DIR/issue_76432.rs:9:54: 9:68
22       let mut _20: *const T;               // in scope 0 at $DIR/issue_76432.rs:9:70: 9:84
23       let mut _21: *const T;               // in scope 0 at $DIR/issue_76432.rs:9:70: 9:84
-       let mut _22: !;                      // in scope 0 at $SRC_DIR/core/src/macros/mod.rs:LL:COL
+       let mut _22: !;                      // in scope 0 at $SRC_DIR/core/src/panic.rs:LL:COL
25       let mut _23: &[T; 3];                // in scope 0 at $DIR/issue_76432.rs:7:19: 7:29
26       scope 1 {
27           debug v => _2;                   // in scope 1 at $DIR/issue_76432.rs:7:9: 7:10
66       }
67   
68       bb1: {
68       bb1: {
-           StorageLive(_22);                // scope 1 at $SRC_DIR/core/src/macros/mod.rs:LL:COL
-           core::panicking::panic(const "internal error: entered unreachable code"); // scope 1 at $SRC_DIR/core/src/macros/mod.rs:LL:COL
+           StorageLive(_22);                // scope 1 at $SRC_DIR/core/src/panic.rs:LL:COL
+           core::panicking::panic(const "internal error: entered unreachable code"); // scope 1 at $SRC_DIR/core/src/panic.rs:LL:COL
71                                            // mir::Constant
-                                            // + span: $SRC_DIR/core/src/macros/mod.rs:LL:COL
+                                            // + span: $SRC_DIR/core/src/panic.rs:LL:COL
73                                            // + literal: Const { ty: fn(&'static str) -> ! {core::panicking::panic}, val: Value(Scalar(<ZST>)) }
74                                            // ty::Const
75                                            // + ty: &str

76                                            // + val: Value(Slice { data: Allocation { bytes: [105, 110, 116, 101, 114, 110, 97, 108, 32, 101, 114, 114, 111, 114, 58, 32, 101, 110, 116, 101, 114, 101, 100, 32, 117, 110, 114, 101, 97, 99, 104, 97, 98, 108, 101, 32, 99, 111, 100, 101], relocations: Relocations(SortedMap { data: [] }), init_mask: InitMask { blocks: [1099511627775], len: Size { raw: 40 } }, align: Align { pow2: 0 }, mutability: Not, extra: () }, start: 0, end: 40 })
77                                            // mir::Constant
-                                            // + span: $SRC_DIR/core/src/macros/mod.rs:LL:COL
+                                            // + span: $SRC_DIR/core/src/panic.rs:LL:COL
79                                            // + literal: Const { ty: &str, val: Value(Slice { data: Allocation { bytes: [105, 110, 116, 101, 114, 110, 97, 108, 32, 101, 114, 114, 111, 114, 58, 32, 101, 110, 116, 101, 114, 101, 100, 32, 117, 110, 114, 101, 97, 99, 104, 97, 98, 108, 101, 32, 99, 111, 100, 101], relocations: Relocations(SortedMap { data: [] }), init_mask: InitMask { blocks: [1099511627775], len: Size { raw: 40 } }, align: Align { pow2: 0 }, mutability: Not, extra: () }, start: 0, end: 40 }) }
81   


thread '[mir-opt] mir-opt/issue_76432.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/issue_76432.test.SimplifyComparisonIntegral.diff', src/tools/compiletest/src/runtest.rs:3372:25


failures:
    [mir-opt] mir-opt/issue_76432.rs
    [mir-opt] mir-opt/issue_76432.rs

test result: FAILED. 160 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 4.22s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/mir-opt" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "mir-opt" "--mode" "mir-opt" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:55
```

</details>

camelid commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 7331ddeed4310ba606bae8479850985251a35dd3 with merge 07d94fb27d02adc112536d7c6439de6a373f2e03...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7331ddeed4310ba606bae8479850985251a35dd3","merge_sha":"07d94fb27d02adc112536d7c6439de6a373f2e03"} -->

camelid commented on src/test/ui/macros/unreachable-2021.rs:
```diff
@@ -0,0 +1,9 @@
+// edition:2021
```
Could you rename this test to something like `unreachable-2021-format_args.rs` to make it clear what it tests?

camelid replied:
And could you also merge this test with the `unreachable-2015` one using [revisions](https://rustc-dev-guide.rust-lang.org/tests/compiletest.html#revisions)?

Urgau replied:
Done in https://github.com/rust-lang/rust/pull/93179/commits/d0e4240a1468b3b2e921324cbb7088850dacb4df

camelid commented:
cc discussion about implementation strategy: https://github.com/rust-lang/rust/issues/92137#issuecomment-1018904497

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4904755531?check_suite_focus=true)
Build commit: 07d94fb27d02adc112536d7c6439de6a373f2e03 (`07d94fb27d02adc112536d7c6439de6a373f2e03`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4904755531?check_suite_focus=true"},"merge_sha":"07d94fb27d02adc112536d7c6439de6a373f2e03"} -->

rust-timer commented:
Queued 07d94fb27d02adc112536d7c6439de6a373f2e03 with parent 17d29dcdce9b9e838635eb0adefd9b8b1588410b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=17d29dcdce9b9e838635eb0adefd9b8b1588410b&end=07d94fb27d02adc112536d7c6439de6a373f2e03).

rust-timer commented:
Finished benchmarking commit (07d94fb27d02adc112536d7c6439de6a373f2e03): [comparison url](https://perf.rust-lang.org/compare.html?start=17d29dcdce9b9e838635eb0adefd9b8b1588410b&end=07d94fb27d02adc112536d7c6439de6a373f2e03).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

m-ou-se commented:
Thanks for working on this! I'll try to review this soon.

I'll have to note that for any edition-specific changes, we also need to add a migration lint to automatically upgrade code from older editions. (Just like we already have for `panic!()`.) Unfortuantely, because Rust 2021 is already stable, this will have to happen at the same time in this case.

> This pull-request also adds an internal feature `format_args_capture_non_literal` that allows capturing arguments from formatted string that expanded from macros. The original RFC #2795 mentioned this as a future possibility. This feature is [required](https://github.com/rust-lang/rust/issues/92137#issuecomment-1018630522) because of concatenation that needs to be done inside the macro:

Another option is to make that work using a nested format_args:

```diff
-        $crate::panic!($crate::concat!("internal error: entered unreachable code: ", $fmt), $($arg)*)
+        $crate::panic!("internal error: entered unreachable code: {}", $crate::format_args!($fmt, $($arg)*))
```

m-ou-se commented on compiler/rustc_builtin_macros/src/unreachable.rs:
```diff
@@ -0,0 +1,63 @@
+use rustc_ast::ptr::P;
```
This file is almost an exact duplicate of `compiler/rustc_builtin_macros/src/panic.rs`. I'd prefer not to duplicate code like this, as it makes maintanance and history tracking more complicated. Instead, we could put both in the same file, and extract the common parts into a common function. E.g. something like:

```rust
pub fn expand_panic<'cx>(cx: &'cx mut ExtCtxt<'_>, sp: Span, tts: TokenStream) -> Box<dyn MacResult + 'cx> {
    let mac = if use_panic_2021(sp) { sym::panic_2021 } else { sym::panic_2015 };
    expand(mac, cx, sp, tts)
}

pub fn expand_unreachable<'cx>(cx: &'cx mut ExtCtxt<'_>, sp: Span, tts: TokenStream) -> Box<dyn MacResult + 'cx> {
    let mac = if use_panic_2021(sp) { sym::unreachable_2021 } else { sym::unreachable_2015 };
    expand(mac, cx, sp, tts)
}

fn expand(mac: Symbol, cx: &'cx mut ExtCtxt<'_>, sp: Span, tts: TokenStream) -> Box<dyn MacResult + 'cx> {
    ...
}
```

Urgau replied:
Done in https://github.com/rust-lang/rust/pull/93179/commits/ac1865fa02e4e2967bed87891ff79be30b9e593e

m-ou-se commented on compiler/rustc_builtin_macros/src/unreachable.rs:
```diff
@@ -0,0 +1,63 @@
+// - `$crate::panic::unreachable_2015!(...)` or
+// - `$crate::panic::unreachable_2021!(...)`
+// depending on the edition.
+//
+// `$crate` will refer to either the `std` or `core` crate depending on which
+// one we're expanding from.
```
This part is not true for `unreachable`, as `std::unreachable` is just a re-export of `core::unreachable`. So `$crate` will always be `core` here.

m-ou-se commented:
> we also need to add a migration lint

I'd be happy to extend the `non_fmt_panics` lint to also cover unreachable. It's not trivial as `panic!(123)` and `unreachable!(123)` are a bit different: the first throws a `i32` as a `Box<dyn Any>`, while the latter just formats the `i32` with `"{}"`. But it shouldn't be too much work.

Urgau commented:
I've updated the implementation to use a nested `format_args!` and reverted the introduction of `format_args_capture_non_literal`. I've also updated and improved the test using revisions as *camelid* [suggested](https://github.com/rust-lang/rust/pull/93179#discussion_r790090480).

> > we also need to add a migration lint
> 
> I'd be happy to extend the `non_fmt_panics` lint to also cover `unreachable`. It's not trivial as `panic!(123)` and `unreachable!(123)` are a bit different: the first throws a `i32` as a `Box<dyn Any>`, while the latter just formats the `i32` with `"{}"`. But it shouldn't be too much work.

I tried to expand the current `non_fmt_panics` but I'm stuck on the fact that the argument passed to `panic_str` is the result of `$crate::concat!("internal error: entered unreachable code: ", $fmt)` and I don't know how (if even possible) to get the argument passed to `concat` or `unreachable_2015`. Any suggestions ?

camelid commented on src/test/ui/macros/unreachable-format-args.rs:
```diff
@@ -0,0 +1,14 @@
+// ignore-emscripten no processes
```
Why are you ignoring emscripten in these tests?

Urgau replied:
I did the same as this test [unreachable-fmt-msg.rs](https://github.com/rust-lang/rust/blob/8196e37ac5c1ed611470faa1f73b4af2f7bc7037/src/test/ui/macros/unreachable-fmt-msg.rs#L3)

m-ou-se commented:
> > 
> 
> I tried to expand the current `non_fmt_panics` but I'm stuck on the fact that the argument passed to `panic_str` is the result of `$crate::concat!("internal error: entered unreachable code: ", $fmt)` and I don't know how (if even possible) to get the argument passed to `concat` or `unreachable_2015`. Any suggestions ?

Maybe you could make the second case of `unreachable_2015` expand to a function call instead: `unreachable_display(&$msg)` and then define a `fn unreachable_display(_: &impl Display) -> !;` to do the panic. Then you can detect that function in the lint, similar to how it detects `panic_any` / `panic_str` / etc.

m-ou-se commented on library/core/src/panic.rs:
```diff
@@ -58,6 +58,42 @@ pub macro panic_2021 {
+    () => (
+        $crate::panicking::panic("internal error: entered unreachable code")
+    ),
+    // Use `panic_str` instead of `panic_display::<&str>` for non_fmt_panic lint.
+    ($msg:expr $(,)?) => (
+        $crate::panicking::panic_str(
+            $crate::concat!("internal error: entered unreachable code: ", $msg))
+    ),
```
Using `concat` rather than formatting with Display breaks code like this:

```rust
    let a = "hello";
    unreachable!(a);
```

Urgau replied:
Fixed with the introduction of `unreachable_display!` and test added https://github.com/rust-lang/rust/pull/93179/commits/9264aa58d2cc38c6a497f19ef616b38c52b3fff6

Urgau committed:
Take in account the unreachable! macro in the non_fmt_panic lint

Urgau commented:
> > I tried to expand the current `non_fmt_panics` but I'm stuck on the fact that the argument passed to `panic_str` is the result of `$crate::concat!("internal error: entered unreachable code: ", $fmt)` and I don't know how (if even possible) to get the argument passed to `concat` or `unreachable_2015`. Any suggestions ?
> 
> Maybe you could make the second case of `unreachable_2015` expand to a function call instead: `unreachable_display(&$msg)` and then define a `fn unreachable_display(_: &impl Display) -> !;` to do the panic. Then you can detect that function in the lint, similar to how it detects `panic_any` / `panic_str` / etc.

Added in https://github.com/rust-lang/rust/pull/93179/commits/8f00aed529488bbb81fd2570167f74a9302b342b and update of `non_fmt_panic` in https://github.com/rust-lang/rust/pull/93179/commits/d8de2383e2c25bd2ad10b44b03646acf1e19247a.

m-ou-se commented on library/core/src/panicking.rs:
```diff
@@ -56,6 +56,16 @@ pub const fn panic_str(expr: &str) -> ! {
+#[cfg(not(bootstrap))]
+#[inline]
+#[track_caller]
+#[lang = "unreachable_display"] // needed for `non-fmt-panics` lint
```
If a lang item is only used for diagnostics, it's better to use `rustc_diagnostic_item` instead.

m-ou-se replied:
(I see that `panic_str` has the same comment. Looks like that's left over from this cleanup: #90273. Originally it wasn't only there for diagnostics. I'll clean up `panic_str` some time soon.)

m-ou-se commented on library/core/src/panicking.rs:
```diff
@@ -56,6 +56,16 @@ pub const fn panic_str(expr: &str) -> ! {
+#[inline]
+#[track_caller]
+#[lang = "unreachable_display"] // needed for `non-fmt-panics` lint
+#[rustc_do_not_const_check] // hooked by const-eval
+#[rustc_const_unstable(feature = "core_panic", issue = "none")]
+pub const fn unreachable_display<T: fmt::Display>(x: &T) -> ! {
```
`unreachable!(arg)` has never worked in const context, and the comment `// hooked by const-eval` is not true, making `rustc_do_not_const_check` problematic. I don't think there's any need to make this function `const`.

m-ou-se commented on library/core/src/panic.rs:
```diff
@@ -58,6 +58,42 @@ pub macro panic_2021 {
+#[allow_internal_unstable(core_panic, const_format_args)]
+#[rustc_diagnostic_item = "unreachable_2021_macro"]
+#[rustc_macro_transparency = "semitransparent"]
+pub macro unreachable_2021 {
+    () => (
+        $crate::panicking::panic("internal error: entered unreachable code")
+    ),
+    ($($t:tt)+) => (
+        $crate::panicking::panic_fmt(
+            $crate::const_format_args!(
+                "internal error: entered unreachable code: {}",
+                    $crate::const_format_args!($($t)+)))
+    ),
+}
```
Let's just use the `panic!()` macro in the second case, to follow the `todo` and `unimplemented` macros: https://github.com/rust-lang/rust/blob/a34c0797528172ede89480e3033f7a5e71ea4735/library/core/src/macros/mod.rs#L688-L691

There's no need for the `const_format_args` here. The regular `format_args` should suffice, as none of this would work in const context anyway.

m-ou-se commented on compiler/rustc_lint/src/non_fmt_panic.rs:
```diff
@@ -67,7 +79,11 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
 }
 
-fn check_panic<'tcx>(cx: &LateContext<'tcx>, f: &'tcx hir::Expr<'tcx>, arg: &'tcx hir::Expr<'tcx>) {
+fn check_panic_or_unreachable<'tcx>(
```
This function also covers `debug_assert` and `assert`, but since `check_panic_or_assert_or_debug_assert_or_unreachable` gets a bit long, I think just `check_panic` suffices, as they're all forms of panic.

m-ou-se commented on library/core/src/panic.rs:
```diff
@@ -58,6 +58,42 @@ pub macro panic_2021 {
 
+#[doc(hidden)]
+#[unstable(feature = "edition_panic", issue = "none", reason = "use unreachable!() instead")]
+#[allow_internal_unstable(core_panic, const_format_args)]
```
```suggestion
#[allow_internal_unstable(core_panic)]
```

m-ou-se commented on compiler/rustc_lint/src/non_fmt_panic.rs:
```diff
@@ -52,13 +52,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
                         ) {
-                            check_panic(cx, f, arg);
+                            check_panic_or_unreachable(cx, f, arg);
+                        } else if matches!(diagnostic_item, Some(sym::unreachable_2015_macro)) {
```
It's probably better to do the removal of the `AddrOf` operator based on the function that was called (`unreachable_display()`), rather than the macro (`unreachable!()`), since `unreachable!()` can also expand to `panic_fn` where this shouldn't happen.

m-ou-se commented on compiler/rustc_lint/src/non_fmt_panic.rs:
```diff
@@ -52,13 +52,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
+                    || Some(def_id) == cx.tcx.lang_items().unreachable_display()
                 {
                     if let Some(id) = f.span.ctxt().outer_expn_data().macro_def_id {
+                        let diagnostic_item = cx.tcx.get_diagnostic_name(id);
                         if matches!(
```
A series of `[else] if matches!()` on the same value can just be written as a regular `match`.

m-ou-se reviewed (changes requested):
Nice! Thanks for working on this!

I have a few more comments:

Urgau commented:
I've address all of review comments. Let me know if there is something else that I should modify or improve.

m-ou-se commented on compiler/rustc_lint/src/non_fmt_panic.rs:
```diff
@@ -49,9 +49,11 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
                 if Some(def_id) == cx.tcx.lang_items().begin_panic_fn()
                     || Some(def_id) == cx.tcx.lang_items().panic_fn()
-                    || Some(def_id) == cx.tcx.lang_items().panic_str()
+                    || matches!(f_diagnostic_name, Some(sym::panic_str))
```
```suggestion
                    || f_diagnostic_name == Some(sym::panic_str)
```

m-ou-se commented on compiler/rustc_lint/src/non_fmt_panic.rs:
```diff
@@ -61,6 +63,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
                             check_panic(cx, f, arg);
                         }
                     }
+                } else if matches!(f_diagnostic_name, Some(sym::unreachable_display)) {
```
```suggestion
                } else if f_diagnostic_name == Some(sym::unreachable_display) {
```

m-ou-se commented on compiler/rustc_lint/src/non_fmt_panic.rs:
```diff
@@ -61,6 +63,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
                     }
+                } else if matches!(f_diagnostic_name, Some(sym::unreachable_display)) {
+                    if let Some(id) = f.span.ctxt().outer_expn_data().macro_def_id {
+                        if matches!(
+                            cx.tcx.get_diagnostic_name(id),
+                            Some(sym::unreachable_2015_macro)
+                        ) {
```
If you're only checking a single diagnostic name + def_id pair, you can use `tcx.is_diagnostic_item(sym, id)` instead.
```suggestion
                        if cx.tcx.is_diagnostic_item(sym::unreachable_2015_macro, id) {
```

m-ou-se reviewed (approved):
Wauw, that was fast.

Other than a few tiny nits, this looks great to me.

Can you squash the commits a bit? There's now a few commits that undo changes of previous commits in this PR, potentially making history tracking and backporting a bit complicated. Ideally this change would be backported to 1.59 (beta) too.

I'd like someone from the compiler team to also quickly look at this, since I'm not on the compiler team. Ping @oli-obk, since you were assigned by high-five.

Let's wait with r+'ing this until the FCP on https://github.com/rust-lang/rust/issues/92137 is finished. We have a few weeks left before the next release cycle.

klensy commented on src/test/ui/consts/const-eval/const_panic.stderr:
```diff
@@ -68,7 +68,7 @@ error[E0080]: evaluation of constant value failed
    |                    ^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'internal error: entered unreachable code', $DIR/const_panic.rs:30:20
    |
-   = note: this error originates in the macro `core::unreachable` (in Nightly builds, run with -Z macro-backtrace for more info)
+   = note: this error originates in the macro `$crate::panic::unreachable_2015` (in Nightly builds, run with -Z macro-backtrace for more info)
```
Previously it was `core::unreachable` and now `$crate::panic::unreachable_2015`, reference to `$crate` as rust library in error message doesn't look good.

m-ou-se replied:
That's not really a problem of this PR. There are already many cases of this for other macros. Here's two examples: https://github.com/rust-lang/rust/blob/1e4067957bd5d0e12c1657e720903209ecc291dc/src/test/ui/consts/const-eval/const_panic.stderr#L7 https://github.com/rust-lang/rust/blob/1e4067957bd5d0e12c1657e720903209ecc291dc/src/test/ui/consts/const-eval/issue-43197.stderr#L42

Might be nice to improve this, but that's not something to block this PR on.

klensy replied:
Oh, didn't notice, but yes, maybe this should be improved.

bors commented:
:umbrella: The latest upstream changes (presumably #93498) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
@bors r=m-ou-se,oli-obk

bors commented:
:pushpin: Commit a4b93eb1883d795f335df71d251d37d17b1311fd has been approved by `m-ou-se,oli-obk`

<!-- @bors r=m-ou-se,oli-obk a4b93eb1883d795f335df71d251d37d17b1311fd -->
<!-- homu: {"type":"Approved","sha":"a4b93eb1883d795f335df71d251d37d17b1311fd","approver":"m-ou-se,oli-obk"} -->

bors commented:
:hourglass: Testing commit a4b93eb1883d795f335df71d251d37d17b1311fd with merge 2b3d81a6965ab4a24d013a7eec98e7369c0edecf...
<!-- homu: {"type":"BuildStarted","head_sha":"a4b93eb1883d795f335df71d251d37d17b1311fd","merge_sha":"2b3d81a6965ab4a24d013a7eec98e7369c0edecf"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

oli-obk commented:
@bors retry timeout

bors commented:
:hourglass: Testing commit a4b93eb1883d795f335df71d251d37d17b1311fd with merge a7ded12529212c61132a0d38be90e80baa965c29...
<!-- homu: {"type":"BuildStarted","head_sha":"a4b93eb1883d795f335df71d251d37d17b1311fd","merge_sha":"a7ded12529212c61132a0d38be90e80baa965c29"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5076845482?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/2b3d81a6965ab4a24d013a7eec98e7369c0edecf/checks/5076845482/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5078370772?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/a7ded12529212c61132a0d38be90e80baa965c29/checks/5078370772/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

camelid commented:
@bors retry timeout

matthiaskrgr commented:
cc https://github.com/rust-lang/rust/issues/93329

bors commented:
:hourglass: Testing commit a4b93eb1883d795f335df71d251d37d17b1311fd with merge 25b21a1d160013b2d58c98d0337c75a721317160...
<!-- homu: {"type":"BuildStarted","head_sha":"a4b93eb1883d795f335df71d251d37d17b1311fd","merge_sha":"25b21a1d160013b2d58c98d0337c75a721317160"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5087882741?check_suite_focus=true)
Approved by: m-ou-se,oli-obk
Pushing 25b21a1d160013b2d58c98d0337c75a721317160 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"m-ou-se,oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5087882741?check_suite_focus=true"},"merge_sha":"25b21a1d160013b2d58c98d0337c75a721317160"} -->

rust-timer commented:
Finished benchmarking commit (25b21a1d160013b2d58c98d0337c75a721317160): [comparison url](https://perf.rust-lang.org/compare.html?start=7b43cfc9b25ac4a906bd56d32d3111085dd9e6a1&end=25b21a1d160013b2d58c98d0337c75a721317160).

**Summary**: This benchmark run shows 11 relevant improvements 🎉 but 1 relevant regression 😿 to instruction counts.
- Average relevant improvement: -1.2%
- Largest improvement in instruction counts: -2.2% on `full` builds of `cranelift-codegen check`
- Largest regression in instruction counts: 1.8% on `incr-patched: b9b3e592dd cherry picked` builds of `style-servo debug`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression

jhpratt commented:
I think this might be the first time I disagree with a perf run. This should be neutral to an improvement, not a regression. I'll leave the issue filing to someone else, though.

camelid commented:
@jhpratt It is weird that the changes are so large, but it's almost entirely an improvement. There's only one regression, which means it's probably spurious.

Urgau commented:
Based on this comment from @m-ou-se:

> Can you squash the commits a bit? There's now a few commits that undo changes of previous commits in this PR, potentially making history tracking and backporting a bit complicated. Ideally this change would be backported to 1.59 (beta) too.

Shouldn't this PR be marked as `beta-nominated` ? or even `stable-nominated` ?

jhpratt commented:
Beta seems reasonable. Not sure we need a point release for this, though.

Urgau commented:
> Beta seems reasonable. Not sure we need a point release for this, though.

Well https://github.com/rust-lang/rust/pull/93394 is in the same wane as this PR and [was](https://github.com/rust-lang/rust/pull/93394#issuecomment-1029229660) `stable-nominated` and `stable-accepted`.

m-ou-se commented:
> Not sure we need a point release for this, though.

As I was recently told, `stable-nominated` doesn't mean it's worth a point release. It means that if a new stable point release happens, this should be included.

I'm not sure if even in that case it should be backported, considering this change is a bit more complex than https://github.com/rust-lang/rust/pull/93394, but it's a discussion that can be had. (But we're getting close to the next release anyway, so a point release seems unlikely at this point.)

Mark-Simulacrum commented:
Marking as perf-regression-triaged -- only regression appears to be due to codegen unit reshuffling, not directly related to this PR.

alice-i-cecile commented:
This broke `bevy_ecs` on nightly. Here was our fix: https://github.com/bevyengine/bevy/pull/3889

Urgau commented:
@alice-i-cecile This is expected, see https://github.com/rust-lang/rust/issues/92137#issuecomment-1019519285 for more context about this change and why the breaking change is considered acceptable. Also note that this change will probably be at least back-ported to beta.

jhpratt commented:
I think I've said this before, but a crater run would be nice to have here. Limited to 2021 edition if it's possible (I don't think it is).

Urgau commented:
> I think I've said this before, but a crater run would be nice to have here. Limited to 2021 edition if it's possible (I don't think it is).

It will be done as part of the crater run done for every new stable (it's actually done with beta but you see the point).

jhpratt commented:
Yeah, I just meant to do it before so we don't have to rush a last-minute revert if it breaks too much.

apiraino commented:
Beta backport accepted as per compiler team [on Zulip](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-02-10.20.2354818.html#271444630)

Stable backport declined ([zulip discussion](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-02-10.20.2354818.html#271448198))

@rustbot label +beta-accepted -stable-nominated
